orderbook/src/database/trades.rs

```diff
@@ -0,0 +1,378 @@
+use crate::conversions::{big_decimal_to_big_uint, h160_from_vec};
+use crate::database::Database;
+use anyhow::{anyhow, Context, Result};
+use bigdecimal::BigDecimal;
+use ethcontract::H160;
+use futures::{stream::TryStreamExt, Stream};
+use model::order::OrderUid;
+use model::trade::Trade;
+use std::convert::TryInto;
+
+/// Any default value means that this field is unfiltered.
+#[derive(Debug, Default)]
+pub struct TradeFilter {
+    pub owner: Option<H160>,
+    pub order_uid: Option<OrderUid>,
+}
+
+impl Database {
+    pub fn trades<'a>(&'a self, filter: &'a TradeFilter) -> impl Stream<Item = Result<Trade>> + 'a {
+        const QUERY: &str = "\
+            SELECT \
+                t.block_number, \
+                t.log_index, \
+                t.order_uid, \
+                t.buy_amount, \
+                t.sell_amount, \
+                t.sell_amount - t.fee_amount as sell_amount_before_fees,\
```

Comment:
Assuming that the smart contract is correct this can't happen though, right? Because sell_amount has already been calculated as fee_amount + traded_amount.